bazaarvoice/emodb
# Modified scheduled stash so it can optionally be run only by request

## Github Issue #

None

## What Are We Doing Here?

Based on our experience there are some environments, such as in a CI or QA environment, where Stash isn't needed every day and as a cost savings it would be better if it didn't run ever day.  One possible change would be to expand the cadence at which Stash could be scheduled (e.g.; every Monday and Thursday).  However, after reviewing with Stash consumers internally there was no single cadence that seemed to fit everyone; the need for Stash in these environments was largely ad-hoc.

This PR introduces the optional ability for Stash to be executed only if explicitly requested by an API call prior to the Stash's scheduled execution time.  For example, if there is a Stash whose ID is "default" then a Stash could be requested using a curl command such as:

```
curl -XPUT stash-host:8080/stash/1/request/default -H "X-BV-API-Key: requesters_api_key"
```

Although there are currently no actions tied to the requesters (such as cost allocation) the requesters are logged in an Emo table for accountability.

Since this PR contains several smaller updates which were necessary for this change here's a list of the most significant changes in this PR:

* As described previously, when Stash starts if the stash only runs by request the run will be skipped if no requests have been made.
* When updating the CloudWatch metric for pending Stashes it will not be updated unless there have been any requests for the Stash.  This only applies to Stashes which run only by request.
* The Stash resource now requires API keys and permissions for all calls.  Previously there were no permission checks.
* Because an ID was required to uniquely identify which Stash the user wanted to run the configuration file was changed slightly such that the "scheduledScan" attribute is now a map of "scheduledScans".  This also expands the configuration to allow for multiple configured scheduled stashes (the backend already supported this but it wasn't configurable).
* Slight refactoring of `EmoModule` was required so the authentication and authorization filters would be applied to the Stash resource.
* In a continuing effort to move away from "scanner" to "stash" nomenclature all new Stash-related files were renamed and followed the "stash" convention.  Existing classes were left unchanged except where doing so led to internal inconsistencies.
* While the above is largely an internal update the externally-visibly API for stash was renamed from "scanner" to "stash" as well.  Since this was an internal interface anyway there shouldn't be any public consequences of this change.
* The Stash resource now registers in ZooKeeper as "emodb-stash-1".  Although this PR does not include a Java Stash client this allows for discovery of Stash by ostrich and allows for a future PR for a client.

## How to Test and Verify

1. As with most Stash updates this is difficult to test locally.  One way to do it is to:
1.1. Run Emo normally using `web-local/start.sh`
1.2. Start Stash using `web-local/start-stash-role.sh`
1.3. Use the Stash API on port 8780 to verify they all work as expected.

## Risk

This is a fairly low risk change.  The Stash process itself doesn't change with this PR, and in those environments where Stash runs daily the process should be unchanged.

### Level 

`Medium`

### Required Testing

`Manual`

Add one or a few complete sentences about the possible risks or concerns for
this change.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
